extensions/amp-story-360/0.1/amp-story-360.js

```diff
@@ -202,22 +266,198 @@ export class AmpStory360 extends AMP.BaseElement {
+    if (permissionState === 'granted') {
+      this.enableGyroscope_();
+    } else if (permissionState === 'denied') {
+      this.gyroscopeControls_ = false;
+      this.togglePermissionClass_(true);
+    }
+  }
+
+  /** @private */
+  checkGyroscopePermissions_() {
+    //  Browser doesn't support DeviceOrientationEvent API.
+    if (typeof DeviceOrientationEvent === 'undefined') {
+      return;
+    }
+
+    // If browser doesn't require permission, enable gyroscope.
+    if (typeof DeviceOrientationEvent.requestPermission === 'undefined') {
+      this.enableGyroscope_();
+    }
+
+    // If DeviceOrientationEvent and permissions, build permission button.
+    if (typeof DeviceOrientationEvent.requestPermission === 'function') {
+      this.buildPermissionButton_();
+    }
+  }
+
+  /**
+   * Creates a device orientation listener and sets gyroscopeControls_ state.
+   * If listener is not called in 1000ms, remove listener and resume animation.
+   * @private
+   */
+  enableGyroscope_() {
+    this.gyroscopeControls_ = true;
+    this.togglePermissionClass_(true);
+
+    this.win.addEventListener('deviceorientation', (e) => {
+      this.isReady_ && this.onDeviceOrientation_(e);
+      clearTimeout(checkNoMotion);
+    });
+
+    const checkNoMotion = setTimeout(() => {
+      this.gyroscopeControls_ = false;
+      this.animate_();
+    }, 1000);
+  }
+
+  /**
+   * @param {Event} e
+   * @private
+   */
+  onDeviceOrientation_(e) {
+    let rot = Matrix.identity(3);
+    rot = Matrix.mul(
+      3,
+      Matrix.rotation(3, 1, 0, (Math.PI / 180.0) * e.alpha),
+      rot
+    );
+    rot = Matrix.mul(
+      3,
+      Matrix.rotation(3, 2, 1, (Math.PI / 180.0) * e.beta),
+      rot
+    );
+    rot = Matrix.mul(
+      3,
+      Matrix.rotation(3, 0, 2, (Math.PI / 180.0) * e.gamma),
+      rot
+    );
+    this.renderer_.setCamera(rot, 1);
+    this.renderer_.render(true);
+  }
+
+  /**
+   * Creates a "activate" button to request DeviceOrientation permissions.
+   * Simulates a click event to call this method on page load.
+   * @private
+   */
+  buildPermissionButton_() {
+    const tempButton = document.createElement('div');
+    tempButton.addEventListener('click', () => {
+      if (this.win.DeviceOrientationEvent.requestPermission) {
+        this.win.DeviceOrientationEvent.requestPermission()
+          // Render activate button if permissions aren't set yet.
+          .catch(() => {
+            const activateButton = buildActivateButtonTemplate(this.element);
+            this.element.appendChild(activateButton);
+
+            activateButton.addEventListener('click', () => {
+              this.requestGyroscopePermissions_();
+            });
+          })
+          // If permissions are already set, handle permission state.
+          .then((permissionState) => {
+            permissionState && this.setPermissionState_(permissionState);
+          });
+      }
+    });
+    tempButton.click();
```

Comment:
Can you explain why you need this `tempButton` and fire a click event right away? It looks like you're trying to detect whether permissions were already granted, but does it work?
I'd be super surprised if Safari doesn't check if the click was programmatic or comes from a user interaction to requestPermission.

Can you ever get to line 384?

Reply:
>Can you explain why you need this tempButton and fire a click event right away? It looks like you're trying to detect whether permissions were already granted, but does it work?

Yes exactly.
We can't call a DeviceOrientationEvent without a user interaction. 
It works on safari and ios chrome.

> Can you ever get to line 384?

This runs on page reload if the user has already set permissions. 

For reference, I got this idea from a Mozilla web vr framework:
https://github.com/aframevr/aframe/blob/7d289e8fb28f2c2895ba49200f502257110d5050/src/components/scene/device-orientation-permission-ui.js#L50

Reply:
Great find, the `requestPermission().catch` makes a lot of sense :))

Can you double check whether the extra programmatic click on `tempButton` makes any difference? I don't know that much about browser but I'd expect us to be able to remove it and things would still work the same.

Reply:
I'm not sure I understand. The click is needed to simulate an interaction so that `DeviceOrientationEvent.requestPermission()` will be called. It doesn't appear to work without it.

Reply:
You're 100% right. It's not needed here. Thank you!